### ISSUE TITLE
Duplicate Node IDs after upgrading to 1.2.3

### ISSUE THREAD

**Issue by mgresser** [hashicorp/consul#4741]
After upgrading to 1.2.3 hosts begin reporting the following when registering with the catalog. The node name for the host making the new reservation and the node name for the conflicted always match.

[ERR] consul: "Catalog.Register" RPC failed to server xxx.xxx.xxx.xxx:8300: rpc error making call: failed inserting node: Error while renaming Node ID: "4833aa15-8428-1d1a-46d8-9dba157dbc60": Node name xxx.xxx.xxx is reserved by node c5dc5b48-f105-79f0-7910-3de6629fddd0 with name xxx.xxx.xxx

Restarting consul on the host seems to resolve the issue at least temporarily. I was able to produce this on a mixed environment consisting of CentOS hosts ranging from major release 5 through 7 but most hosts on CentOS 7.

**Comment by pierresouchay** [comment]
Did your node was cleaned up and re-installed? (I mean is the message possible?)

Do you use the option leave on terminate? 

Do you use https://www.consul.io/docs/agent/options.html?#disable_host_node_id with false value (do you have fixed ID per node - I assume you are not)?

**Comment by mgresser** [comment]
I upgraded from 1.2.2 to 1.2.3 so the consul agent restarted for that upgrade. I don't have the leave on terminate or disable_host_node_id options on. I've upgraded many times with my current settings without issue.

**Comment by JackXu2** [comment]
yes, i just install 1.2.3, when the node2 restart, also has this issue 

    2018/10/03 23:12:28 [WARN] consul.fsm: EnsureRegistration failed: failed inserting node: Error while renaming Node ID: "af57f171-44ed-b905-1452-8dfc775afb40": Node name node2 is reserved by node cc5f253c-7498-4d8b-519b-6ad213139023 with name node2
    2018/10/03 23:12:28 [WARN] consul.fsm: EnsureRegistration failed: failed inserting node: Error while renaming Node ID: "af57f171-44ed-b905-1452-8dfc775afb40": Node name node2 is reserved by node cc5f253c-7498-4d8b-519b-6ad213139023 with name node2
    2018/10/03 23:12:28 [WARN] consul.fsm: EnsureRegistration failed: failed inserting node: Error while renaming Node ID: "af57f171-44ed-b905-1452-8dfc775afb40": Node name node2 is reserved by node cc5f253c-7498-4d8b-519b-6ad213139023 with name node2
    2018/10/03 23:12:28 [WARN] consul.fsm: EnsureRegistration failed: failed inserting node: Error while renaming Node ID: "af57f171-44ed-b905-1452-8dfc775afb40": Node name node2 is reserved by node cc5f253c-7498-4d8b-519b-6ad213139023 with name node2
    2018/10/03 23:12:28 [WARN] consul.fsm: EnsureRegistration failed: failed inserting node: Error while renaming Node ID: "af57f171-44ed-b905-1452-8dfc775afb40": Node name node2 is reserved by node cc5f253c-7498-4d8b-519b-6ad213139023 with name node2
    2018/10/03 23:12:28 [WARN] consul.fsm: EnsureRegistration failed: failed inserting node: Error while renaming Node ID: "af57f171-44ed-b905-1452-8dfc775afb40": Node name node2 is reserved by node cc5f253c-7498-4d8b-519b-6ad213139023 with name node2
    2018/10/03 23:12:28 [WARN] consul.fsm: EnsureRegistration failed: failed inserting node: Error while renaming Node ID: "af57f171-44ed-b905-1452-8dfc775afb40": Node name node2 is reserved by node cc5f253c-7498-4d8b-519b-6ad213139023 with name node2
    2018/10/03 23:12:28 [WARN] consul.fsm: EnsureRegistration failed: failed inserting node: Error while renaming Node ID: "af57f171-44ed-b905-1452-8dfc775afb40": Node name node2 is reserved by node cc5f253c-7498-4d8b-519b-6ad213139023 with name node2
    2018/10/03 23:12:28 [WARN] consul.fsm: EnsureRegistration failed: failed inserting node: Error while renaming Node ID: "af57f171-44ed-b905-1452-8dfc775afb40": Node name node2 is reserved by node cc5f253c-7498-4d8b-519b-6ad213139023 with name node2
    2018/10/03 23:12:28 [WARN] consul.fsm: EnsureRegistration failed: failed inserting node: Error while renaming Node ID: "af57f171-44ed-b905-1452-8dfc775afb40": Node name node2 is reserved by node cc5f253c-7498-4d8b-519b-6ad213139023 with name node2

**Comment by pierresouchay** [comment]
Hello @JackXu2 & @mgresser 

Actually, we also had this issue for a precise reason: we re-install nodes from scratch. When re-installing a node - starting with version 0.8.5 - the node id is no longer predictable by default - thus, the nodeid might be different, which leads to this message. We fixed this (on bare-metal hosts) by forcing the predictible nodeid per host, thus, when we re-install the node, Consul sees it as the same machine. Another way is to properly leave the cluster when the "first" node is being decommissioned.

The other possibility to trigger this is when using VMs, if you re-run a node with an identical node name, its ID is also different, thus, Consul refuses the new node to "steal" the node name as the previous node was seen not that much time ago.

Since I worked on this feature (since we had 2 nodes stealing their names) in https://github.com/hashicorp/consul/pull/4415, I am quite interested by your specific usage: are you in one of those cases or is it another corner case?

**Comment by davidkarlsen** [comment]
I'm seeing the same - we're running with the docker image from docker-hub. We set the node name fixed with the -node option.

**Comment by planetxpress** [comment]
We are also having to downgrade back to 1.2.2 as this preventing us from re-deploying clients.

It is a very common pattern to destroy and re-create instances when re-deploying clients which could have the same hostname. In any case they are fresh instances and should be considered as new nodes and not try to inherit anything the cluster may have known about the previous node as they will have a new IP address, so we do not set deterministic node ID. We do have force_leave_on_terminate set however instances always getting a clean shutdown before they are deleted is not a guaranteed thing in all cloud providers.

**Comment by planetxpress** [comment]
Going to amend my previous comment on this after I did some extensive testing on the issue we are now hitting with 1.2.3, and if there is a suggested config path to avoid this issue for this version and beyond it would be much appreciated.

Our consul clients are typically Linux hosts in various cloud providers (Openstack, AWS, GCP), and while we typically re-use hostnames when we delete and re-create clients, they will typically get a new random IP address from the subnet. We have disable_host_node_id at default (true) as we can't use deterministic node ID's due to these hosts getting new IP addresses. We also have leave_on_terminate=true configured for our consul clients.

If a client gets a clean shutdown, we have no problem, because it leaves the cluster. If a client does not get a clean shutdown and remains registered in the cluster, but is rebuilt with a different IP, we also have no problem. The problem happens if a client does not get a clean shutdown, remains registered in the cluster, and is rebuilt with the same IP. Now we run into the node renaming error.

Since we can't guarantee a clean shutdown, nor guarantee that the IP will be different when rebuilding a host in an environment where we can't use deterministic node IDs, what would be the recommended configuration path in this case?

**Comment by planetxpress** [comment]
I was able to duplicate this issue on a node that did not get a clean shutdown and was rebuilt with a different IP address, so it does not seem like the IP it gets actually matters in this case, only that it has the same hostname.

At this point we aren't sure the proper way to handle consul clients that don't get a clean shutdown that need to be re-created. Shouldn't this be a common expected pattern of deployment and not require manual intervention of forcing clients to leave?

**Comment by pierresouchay** [comment]
@planetxpress well, on our side, we are using baremetal nodes with fixed node-ID generation.  

It means that when the node is renamed, it gets its new name, but this protects from having 2 nodes registering with the same name (which did broke our production twice).

When we replace a node by another one, we also give it another name, so not a real issue for us. For users that wish to reuse names quickly for different machines with different IDs not properly leaving the cluster before shutting down, it might create troubles however.

The solution might to be able to skip this test using a new configuration option. @mkeeler do you have an opinion on this?

**Comment by rafaelmagu** [comment]
So, to confirm, there's _no_ fix for an existing node in a cloud environment that runs into this issue?

**Comment by pierresouchay** [comment]
There are actually several issues:
 - I discovered than when not run as root, Consul can not have a stable ID, even when using `"disable_host_node_id": false` => when re-installing a node, its ID will change, hopefully fixed by https://github.com/shirou/gopsutil/pull/603
 - for those not using this flag, we should find something more clever.

I was thinking about adding an option for instance `"allow_node_overwrite": true|false` that would: allow renaming only when existing node is effectively DOWN (meaning, its Serf Health Status is DOWN). Using this option would keep the existing behaviour, but would allow to quickly replace a node with a node with similar name, but only if previous node is already dead (very useful for Cloud based workloads).

Still, I am wondering why `disable_host_node_id` was set to true... because it was not working or another reason?

@banks @mkeeler What do you think?

**Comment by TylerLubeck** [comment]
Any advice on downgrading to 1.2.2? When we do so, we get errors like
```
raft: Failed to restore snapshot 13931-7840952-1542219582505: Unrecognized msg type 15
```

I think this is because [#4535](https://github.com/hashicorp/consul/pull/4535/files#diff-a520cb0eb589c47197ba1ffa7d044bcaR49) landed in 1.2.3 as well.

**Comment by banks** [comment]
> I think this is because #4535 landed in 1.2.3 as well.

@TylerLubeck you are exactly right. Unfortunately downgrading is not easy when there have been new messages added to the state store that the older version can't decode.

In some rare cases where we really know it can't be a problem we do have a way to mark some message types we add as "safe to ignore" which means older versions can downgrade.

In this specific case that wasn't possible because the change in 1.2.3 was to fix a bug where we forgot to store some pretty critical bits of state in the snapshot!

We can't make it safe to downgrade automatically in those cases because the state could become inconsistent if some new objects that were written get dropped and lost but other state that was written after or in same transaction is left.

This is why we recommend taking a snapshot before any upgrade so if you do need to revert you can use that, but I appreciate it's not always possible especially if you don't find the issue that needs a downgrade until much later.

Overall to others, we have this issue on our radar and it highlights how tricky some of the reasoning around "host identity" is. 

The hard thing is trying to reconcile all of these common patterns - initially Consul was closely tied to hostnames and IPs alone and that caused problems for folks running on containers or VMs where the IP could change. It also caused problems for anyone who had to rename their hosts without changing IP.

We introduced IDs to try and get around this but then there is no good way to unique identify a host - the closest are the hardware identifiers or state on disk (how we manage node ids now) but then that changes if you VM dies and gets replaces - kinda exactly what it's meant to do.

It feels a bit like we fix a "bug" with one use case it causes other legitimate use-cases trouble in ways that were hard to anticipate (e.g. only if you fail to leave gracefully.

That said, there is clearly an issue here to resolve (or a few) so we will consider it more.

It might be that we just make the node name stealing protection configurable like Pierre suggested but that would likely come with caveats.

**Comment by pierresouchay** [comment]
@lanetxpress @rafaelmagu what do you think about my proposal in #5008 ?

**Comment by planetxpress** [comment]
@pierresouchay thank you having that option exposed would be fantastic.

**Comment by rafaelmagu** [comment]
@pierresouchay that would work for me as well.

**Comment by linydquantil** [comment]
how to avoid this problem temporarily?

**Comment by pierresouchay** [comment]
`consul leave ; service consul restart` does the trick when we have this issue.

Other way, as we did (for bare-metal servers), compute the node-id with a fixed ID for nodes. We are doing this with the output of dmidecode on our side this way (which is the same way as Consul was supposed to work on Linux): https://github.com/hashicorp/consul/issues/4914#issuecomment-439414989

If you want something very simple, compute the nodeID using the hostname, example (in Ruby):

```ruby
#!/usr/bin/env ruby
require 'digest'

def generate_node_id_for_consul(baseid)
  h = Digest::SHA512.hexdigest(baseid)
  "#{h[0..7]}-#{h[8..11]}-#{h[12..15]}-#{h[16..19]}-#{h[20..31]}"
end

nodeid=generate_node_id_for_consul(`hostname -f`.strip.downcase)

if ARGV.count > 0
  dst=ARGV[0]
  begin
    file = File.open(dst, "w")
    file.write(nodeid)
  rescue StandardError => e
    STDERR.puts "[ERROR] Cannot write to #{dst}: #{e}"
    exit 1
  ensure
    file.close unless file.nil?
  end
else
  puts nodeid
end
exit 0
```

Copy this ruby script somewhere and call it that way before starting consul:
```shell
/my/path/to/script/node-id-gen.rb /var/lib/consul/node-id
```

=> it will generate a predictable `/var/lib/consul/node-id` file base on the FQDN of your machine

**Comment by linydquantil** [comment]
Got it, thanks

**Comment by linydquantil** [comment]
my consul server cluster builds on https://github.com/kelseyhightower/consul-on-kubernetes ; At the same time, an consul client service was launched inside the pod of each project to provide functions such as service registration;  
Excuse me, is this method correct?
My usual yaml file is as follows：
···
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ${project}${domain}
  labels:
    app: ${project}${domain}
spec:
  replicas: 1
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 1
  selector:
    matchLabels:
      app: ${project}${domain}
  template:
    metadata:
      annotations:
      labels:
        app: ${project}${domain}
    spec:
      imagePullSecrets:
        - name: myregistrykey
      containers:
        - name: consul-client
          image: "consul:1.4.0"
          imagePullPolicy: IfNotPresent
          args:
            - "agent"
            - "-advertise=$(PODIP)"
            - "-bind=0.0.0.0"
            - "-retry-join=consul-0.consul.$(NAMESPACE).svc.cluster.local"
            - "-retry-join=consul-1.consul.$(NAMESPACE).svc.cluster.local"
            - "-retry-join=consul-2.consul.$(NAMESPACE).svc.cluster.local"
            - "-client=0.0.0.0"
            - "-encrypt=$(GOSSIP_ENCRYPTION_KEY)"
            - "-datacenter=$(DC)"
            - "-data-dir=/consul/data"
          env:
            - name: PODIP
              valueFrom:
                fieldRef:
                  fieldPath: status.podIP
            - name: NAMESPACE
              valueFrom:
                fieldRef:
                  fieldPath: metadata.namespace
            - name: GOSSIP_ENCRYPTION_KEY
              value: "xxxxx"
            - name: DC
              value: ${host}
          lifecycle:
            preStop:
              exec:
                command:
                - /bin/sh
                - -c
                - consul leave
          resources:
            limits:
              cpu: "200m"
              memory: 512Mi
            requests:
              cpu: "100m"
              memory: 256Mi
          ports:
            - containerPort: 8500
              name: ui-port
            - containerPort: 8400
              name: alt-port
            - containerPort: 53
              name: udp-port
            - containerPort: 8443
              name: https-port
            - containerPort: 8080
              name: http-port
            - containerPort: 8301
              name: serflan
            - containerPort: 8302
              name: serfwan
            - containerPort: 8600
              name: consuldns
            - containerPort: 8300
              name: server
        - name: ${project}${domain}
          image: xxxx/${project_type}/${project}:${version}
          imagePullPolicy: Always
          securityContext:
            capabilities:
              add:
              - SYS_MODULE
              - NET_ADMIN
              - SYS_ADMIN
          env:
            - name: CONFIG_HASH
              value: ${config_hash}
            - name: NODE
              value: ${host}
            - name: PODIP
              valueFrom:
                fieldRef:
                  fieldPath: status.podIP
          resources:
            limits:
              cpu: "2"
              memory: 6Gi
            requests:
              cpu: "2"
              memory: 6Gi
          ports:
            - containerPort: 8080
              name: app-port
···

**Comment by pierresouchay** [comment]
Here is a more elaborated script to generate node-ids: https://github.com/pierresouchay/consul-ops-tools/blob/master/bin/consul_node_id.rb

Example of usage:
```bash
sudo ./consul_node_id.rb -c "hostname -f" -w /var/lib/consul/nodeid
```

**Comment by linydquantil** [comment]
This problem has become more and more frequent recently. Can i just rollback to 1.2.2 ?? 
ps: my consul server cluster builds on https://github.com/kelseyhightower/consul-on-kubernetes

**Comment by pierresouchay** [comment]
@linydquantil You might vote for https://github.com/hashicorp/consul/pull/5008 ... about downgrading... the problem is in the servers, so, if you want to do that, you have to downgrade agents first... might be tricky since downgrade is not supported AFAIK for servers

**Comment by linydquantil** [comment]
ok thanks @pierresouchay

**Comment by aashitvyas** [comment]
We are doing upgrade of the consul from version 1.0.6 to 1.3.1 and encountering the similar issues. We have not tried the option "-disable-host-node-id=false" in our consul config file since it was just upgrade and it should have worked. I am curious to know what our options are in this case ? Should we run the consul as a root and try the "-disable-host-node-id=false" or should we try to upgrade to the version where this issue is not present at the moment ? [ I am curious to know what that version is actually !]

**Comment by aashitvyas** [comment]
Actually from the comment @mgresser , it was working fine until version 1.2.2 but got broke on 1.2.3. We will try to upgrade until 1.2.2 and see if the issue still persists

**Comment by mgresser** [comment]
I've been running with "-disable-host-node-id=false" for a week or so now on version 1.2.2 and it seems to be working as expected thus far. I'm considering attempting this upgrade again with that config change in place next week. I'll update on whether that helps me get onto 1.4+

**Comment by aashitvyas** [comment]
Thank you @mgresser for the follow up. I am curious to know ,  where does it say that the fix will be merged with next week or so ? If it will , then I will wait and try to do the upgrade from 1.0.6 to 1.3.1 with the fix.

**Comment by mgresser** [comment]
I don't know of any ETA for a fix. I'm just going to attempt the upgrade again with the "-disable-host-node-id=false" flag set in my cluster to see if that resolves the issue and allows me to upgrade my hosts.

**Comment by pierresouchay** [comment]
@mgresser Until https://github.com/hashicorp/consul/pull/4926 is merged, `-disable-host-node-id=false` has no real effect, because the value changes at each boot (so, in case of a re-install, it will have a new node ID unless you are running Consul as root).

I would recommend you to use https://github.com/pierresouchay/consul-ops-tools/blob/master/bin/consul_node_id.rb if you need constant IDs between re-installs

**Comment by aashitvyas** [comment]
@pierresouchay  Not sure if you want me to open the separate issue to track this , but I am upgrading from 1.0.6 to 1.3.1 and encountering the same errors on AWS linux instances

**Comment by pierresouchay** [comment]
@aashitvyas for this either multiple policies from #5008, either generate a node ID dependant of your node Name, for instance, when booting:

consul_node_id.rb -c hostname /var/lib/consul/node-id

**Comment by mgresser** [comment]
@pierresouchay Thanks, I hadn't noticed that caveat. That'll save me some effort. I'll wait for the release with the fix.

**Comment by aashitvyas** [comment]
Thank you! @pierresouchay  Do we have any ETA for #5008  to be merged?

**Comment by pierresouchay** [comment]
@aashitvyas Vote for it, but I am not from Hashicorp, I cannot decide :)

**Comment by aashitvyas** [comment]
@pierresouchay oops !! sorry didn't know, I am  newbie on Github. I will vote for it to get some attention

**Comment by EugenMayer** [comment]
Same issue here, specifically when upgrading from 1.2.2 to 1.2.3 and also using `-disable-host-node-id=false` - in my case it happens when i stop/rm a docker container ( service ) and then readd it again, while the consul server keeps it's state.

**Comment by AlexanderZagaevskiy** [comment]
Hi, everyone! We've also bumped into the issue. But what was strange: every next trying to register a client had finished with an error, that node name was reserved by a previously created client.

Here's filtered logs from a normally working client:
```
==> Starting Consul agent...
==> Consul agent running!
           Version: 'bf500a2'
           Node ID: 'b1431f6d-5337-27e1-b045-4ed8615642af'
         Node name: 'QA-T49'
        Datacenter: 'primary' (Segment: '')
...
    2019/01/30 13:48:58 [DEBUG] agent: Using random ID "b1431f6d-5337-27e1-b045-4ed8615642af" as node ID
```
But after some problems with cluster state, next client's launches from this host ended up with errors:
```
==> Starting Consul agent...
==> Consul agent running!
           Version: 'bf500a2'
           Node ID: '4683b4c4-e50d-63f0-efca-ff4e5da9b0e6'
         Node name: 'QA-T49'
        Datacenter: 'primary' (Segment: '')
...
    2019/01/30 13:51:09 [DEBUG] agent: Using random ID "4683b4c4-e50d-63f0-efca-ff4e5da9b0e6" as node ID
    2019/01/30 13:51:12 [ERR] consul: "Catalog.Register" RPC failed to server 172.17.18.40:8300: rpc error making call: failed inserting node: Error while renaming Node ID: "4683b4c4-e50d-63f0-efca-ff4e5da9b0e6": Node name QA-T49 is reserved by node b1431f6d-5337-27e1-b045-4ed8615642af with name QA-T49
...

==> Starting Consul agent...
==> Consul agent running!
           Version: 'bf500a2'
           Node ID: '2b4f6ba0-20ff-2557-30cf-444d8460fb68'
         Node name: 'QA-T49'
        Datacenter: 'primary' (Segment: '')
...
    2019/01/30 14:21:56 [DEBUG] agent: Using random ID "2b4f6ba0-20ff-2557-30cf-444d8460fb68" as node ID
    2019/01/30 14:21:56 [ERR] consul: "Catalog.Register" RPC failed to server 172.17.9.30:8300: rpc error making call: rpc error making call: failed inserting node: Error while renaming Node ID: "2b4f6ba0-20ff-2557-30cf-444d8460fb68": Node name QA-T49 is reserved by node 4683b4c4-e50d-63f0-efca-ff4e5da9b0e6 with name QA-T49
```
... and so on. As you could note, next node could not register because of the previous one.

**Comment by pierresouchay** [comment]
@AlexanderZagaevskiy yes, that's the same issue. You can either vote for 
https://github.com/hashicorp/consul/pull/5008 or simply generate a fixed node ID as explained here: 
https://github.com/hashicorp/consul/issues/4741#issuecomment-442657088 and https://github.com/hashicorp/consul/issues/4741#issuecomment-442749683

**Comment by pearkes** [comment]
We're likely going to address this by utilizing Serf, but will also consider the full discussion in #5008. See this comment for more on that approach: https://github.com/hashicorp/consul/pull/5008#issuecomment-447504127

**Comment by koshatul** [comment]
I use this to generate the node-id to keep the node-ids the same across reboots and rebuilds.
```
uuidgen -N "$(hostname -s)" --namespace "@dns" --sha1
```

**Comment by avoidik** [comment]
may I ask for the status? has it been addressed with https://github.com/hashicorp/consul/pull/4415 ?

**Comment by wimfabri** [comment]
to generate a nodeid with ansible, I have this in the template for the config file:
  "node_id": "{{ ansible_hostname | to_uuid }}",

**Comment by deeco** [comment]
I am seeing this issue with 1.4.4, instance is destroyed without deregitering, when spin up again with same name and ip I get below,  I am testing this in lab and new enough to consul setup.


`2019/04/19 14:24:04 [ERR] agent: failed to sync remote state: rpc error making call: rpc error making call: failed inserting node: Error while renaming Node ID: "e1ba7d76-2da8-46ae-cc32-9b54c7713a01": Node name consultest is reserved by node e48aca4a-49d7-9207-744b-4224dcd25916 with name consultest`

**Comment by banks** [comment]
@avoidik status is that #5485 should fix this which is blocked on https://github.com/hashicorp/memberlist/pull/189 which is blocked un us all racing to get other stuff done for a release deadline!

So the fix is in sight but just needs last few tweaks and merges. I suspect it will be fixed in a patch release after 1.5.0, probably 1.5.1.
